File: Python/recursion/minimum_coins.py
import sys


def min_coins(coins, target):
    if target == 0:
        return 0
    res = sys.maxsize
    for coin in coins:
        if coin <= target:
            rem_target = 1 + min_coins(coins, target - coin)
            if rem_target != sys.maxsize:
                res = min(res, rem_target)
    return res


if __name__ == "__main__":
    print(min_coins([10, 25, 5], 30))

File: Java/recursion/MinimumCoins.java
package recursion;

public class MinimumCoins {
  public int minCoins(int[] coins, int target) {
    if (target == 0)
      return 0;
    int res = Integer.MAX_VALUE;
    for (int coin : coins) {
      if (coin <= target) {
        int remTarget = 1 + minCoins(coins, target - coin);
        if (remTarget != Integer.MAX_VALUE)
          res = Math.min(res, remTarget);
      }
    }
    return res;
  }

  public static void main(String[] args) {
    MinimumCoins ob = new MinimumCoins();
    System.out.println(ob.minCoins(new int[] { 10, 25, 5 }, 30));
  }
}
